rumqttd/src/link/remote.rs

```diff
@@ -225,3 +210,157 @@ where
     // Ok((connect, props, lastwill, lastwill_props))
     Ok(packet)
 }
+
+fn handle_auth(
+    config: Arc<ConnectionSettings>,
+    login: Option<&Login>,
+    client_id: String,
+) -> Result<(), Error> {
+    if config.static_auth.is_none() && config.dynamic_auth.is_none() {
+        return Ok(());
+    }
+
+    if login.is_none() && (config.static_auth.is_some() || config.dynamic_auth.is_some()) {
+        return Err(Error::InvalidAuth);
+    }
+
+    let login = login.unwrap();
+    let u = login.username.as_str();
+    let p = login.password.as_str();
+
+    let mut static_auth_verified = false;
+    if let Some(ref pairs) = config.static_auth {
+        for (validu, validp) in pairs {
+            if u == validu && p == validp {
+                static_auth_verified = true;
+                break;
+            }
+        }
+    }
+
+    if static_auth_verified {
+        return Ok(());
+    }
+
+    if let Some(ref auth) = config.dynamic_auth {
+        if auth(client_id, u.to_owned(), p.to_owned()) {
+            return Ok(());
+        }
+    }
```

Comment:
some points / questions:
- we should check for either static or dynamic, not both!
- dynamic auth should take preference, if it is present, we don't check for static auth.

so basically, if user adds dynamic auth, the auth from config will get ignored.